src/main/java/org/elasticsearch/search/aggregations/bucket/BestDocsDeferringCollector.java

```diff
@@ -225,15 +299,30 @@ public long cost() {
             throw new ElasticsearchException("This caching scorer implementation only implements score() and docID()");
         }
 
-        public void collect(int docId) throws IOException {
-            currentLeafCollector.collect(docId);
+        public void collect(int docId, long parentBucket) throws IOException {
+            perBucketSamples = bigArrays.grow(perBucketSamples, parentBucket + 1);
+            PerParentBucketSamples sampler = perBucketSamples.get((int) parentBucket);
+            if (sampler == null) {
+                sampler = new PerParentBucketSamples(parentBucket, currentScorer, readerContext);
+                perBucketSamples.set((int) parentBucket, sampler);
+            }
+            sampler.collect(docId);
             maxDocId = Math.max(maxDocId, docId);
         }
     }
 
 
-    public int getDocCount() {
-        return matchedDocs;
+    public int getDocCount(long parentBucket) {
+        PerParentBucketSamples sampler = perBucketSamples.get((int) parentBucket);
+        if (sampler == null) {
+            return 0;
```

Comment:
Should we throw an exception here instead? It seems like something has gone wrong if we ask for the doc count for a parentBucket that doesn't exist?